Unify the representation of directories into a single node type

jmmv opened the pull request:
The previous split of MappedDir vs. ScaffoldDir was very complex and got
in the way of upcoming changes to make reconfigurations less intrusive.

Merging the implementation of ScaffoldDir into MappedDir shows that we
can represent a directory in a much simpler way: a way that has a single
mapping of directory entries to nodes, not a layered implementation with
three different maps.

Granted: the new representation is less flexible than the previous one
and requires mappings to be provided in order by the user, but this
restriction feels natural anyway and the gained simplicity is worth it.

Note that, while ScaffoldDir is gone, I'm leaving MappedDir's name
unmodified to keep the churn low in this change.  I'll get rid of the
"mapped" prefix later on.

jmmv commented on internal/sandbox/sandbox_test.go:
```diff
@@ -0,0 +1,290 @@
+//
+// The path argument represents the current location of the got and want nodes relative to the root
+// of the hierarchy.  This value is used purely for error reporting purposes.
+func compareTrees(path string, got Node, want Node) error {
```
Done.

jmmv commented on internal/sandbox/sandbox_test.go:
```diff
@@ -0,0 +1,290 @@
+	}
+	defer os.RemoveAll(tempDir)
+
+	tempDir2 := filepath.Join(tempDir, "nested-dir")
```
Renamed the other two variables as well to make them consistent.

jmmv commented on internal/sandbox/sandbox_test.go:
```diff
@@ -0,0 +1,290 @@
+				return fmt.Errorf("in %s: missing entry %s", path, wantName)
+			}
+		}
+	}
```
I don't think so. The only thing to check in the other case is that the types of the entries match, but we have already done that at the beginning of the function.

jmmv commented on internal/sandbox/sandbox.go:
```diff
@@ -155,45 +153,68 @@ func Serve(c *fuse.Conn, node Dir, dynamic *DynamicConf) error {
-		return len(mappings[i].Mapping) > len(mappings[j].Mapping)
-	})
+	tokens := []string{}
+	lastI := 0
```
Done.

jmmv commented:
I have no idea why I was not allowed to "start a review" and instead had to reply to your individual comments separately.

Rebased the branch on top of the Bazel changes.

Remember that this PR contains three individual commits. Reviewing them separately should make the process much simpler.

rupertks commented:
I haven't actually reviewed the whole thing yet!

If I click each commit, it looked like I could start a separate review for each commit, and approve each commit separately, but apparently not. I guess you really have to make three separate reviews for that process to work. Anyway, still reading...

rupertks commented on internal/sandbox/mapped_dir.go:
```diff
@@ -77,142 +80,196 @@ type openMappedDir struct {
-	for name, node := range o.dir.scaffoldDirs {
-		if !done[name] {
+
+	dirEntries := make([]fuse.Dirent, 0, len(o.dir.children)+len(dirents))
```
dirents vs dirEntries is a bit confusing here

How about renaming readEntries() to readNativeEntries()

And renaming dirents to nativeDirEntries

And renaming dirEntries to fuseDirEntries

Or something like that... ? I'm not sure, probably you can think of better names :)

jmmv replied:
Done.

jmmv commented:
I can send you three individual pull requests if that will shorten the turnaround time.

But I kinda like a pull request with multiple changes in it: these three changes are related and towards one goal. Merging the PR at once (not rebasing, not squashing) keeps that concept in version control because the history shows the grouping.

Up to you.

rupertks commented:
Thanks! 

Sorry about the delay. I think you should keep this one with the multiple commits. But smaller pull requests would definitely make it easier to review things in the future.

jmmv commented:
I'm going to wait until the review is done before applying cosmetic changes... addressing your comments in chunks is just going to make your life more difficult thanks to GitHub's poor review system. Please let me know.

rupertks commented:
Thanks!

I don't have any other comments right now, but one of the reasons that large changes are hard to review in GitHub is that the UI seems to have no memory of collapsed file diffs. 

So every time I revisit this pull request, there are e.g. 24 expanded diffs to look at (well, 23, because GitHub don't expand the mapped_dir.go diff automatically because it thinks it is too big).

Anyway, I don't think delaying an update will help with that. Even if I just refresh the page then everything gets expanded again. But is there a trick that I am missing?

jmmv commented:
I don't know. I haven't used GitHub's review system much... but so far my experiences have been pretty awful. Maybe Philipp knows of some tricks...

rupertks commented:
I'm not sure if we should be using GitHub in a different way. Maybe it would make reviews easier if we added more commits during the review, instead of modifying the original commits?

e.g. There is a new "Unify the representation of directories into a single node type 5b9f460" link above, which I'm assuming means you just changed something, but if I click the link then it shows me the entire diff vs master, instead of showing the diff since the last review/last comment. 

I'm assuming the only thing that changed just then relates to dirEntries, but there is no way for me to know without reading the whole diff again, and even then I wouldn't really know what changed, since I don't have a perfect memory :)

jmmv commented:
That's what I mentioned in the past and tried at some point, but I can't tell if it made a difference.

The other alternative is to send pull requests with just one commit in them. But that seems to go against Git's graph model by linearizing all reviews and commits.

philwo commented:
> I'm not sure if we should be using GitHub in a different way. Maybe it would make reviews easier if we added more commits during the review, instead of modifying the original commits?

I'm pretty sure GitHub pull requests are meant to be added to and never force pushed. Force pushing after modifying seems to confuse the system.

That's also why (IMHO) you shouldn't use the normal "merge" for pull requests to get them in, but instead use ["squash & merge"](https://help.github.com/articles/about-pull-request-merges/#squash-and-merge-your-pull-request-commits). This gives you a clean history and gets rid of the "Fix a typo in a comment in an earlier commit of this pull request" commits that are no longer relevant once a PR is fully reviewed.

jmmv commented:
That assumes you want to squash all commits, which makes sense for the "Fix tyop" garbage that results from a code review but makes no sense here.

I explicitly created a branch to group related commits. Reaching a semantical goal often requires many steps, and it's good to capture those in the version history. This is why this PR had 3 commits. This is why I'll merge it using a merge commit, not a fast-forward. This is how distributed version control systems are supposed to be used.

I could instead have added individual extra commits to address review comments and only squashed those right before merging. It's painful though, but avoids the force-push until the very last minute.

Or I could have sent each commit as a separate PR. But this goes against how DVCS work per the comment above: submitting each change separately and merging it to head with a fast-forward essentially loses track of the relation between the commits.

philwo commented:
> I explicitly created a branch to group related commits. Reaching a semantical goal often requires many steps, and it's good to capture those in the version history. This is why this PR had 3 commits. This is why I'll merge it using a merge commit, not a fast-forward. This is how distributed version control systems are supposed to be used.

I totally agree, but this is not (well) supported by GitHub's code review workflow and is the main complaint I know about it vs. Gerrit. GitHub's workflow doesn't support "commit chains" well, instead it has the "one PR is one semantic change + follow-up commits that address code review comments" model.

I think it's best not to argue in terms of how DVCS are "supposed to work" and just accept that GitHub is a more centralized approach and thus works differently, but is IMHO also not too bad.
